Dozuki/CloudPrem-Infra
# Webhooks & Other Improvements

# Background
Frontegg webhooks were added to the application and deployed on prod. While this solution would technically work in cloudprem, it would expose our customer's webhook data to a third party vendor. Usually our customers opt for cloudprem to keep tight control of their data and this would be the antithesis of that. 

# Solution
To support the data protection our clients expect we had to deploy additional infrastructure to effectively mimic frontegg's service in our individual sequestered cloudprem installations. They call this "hybrid mode". It requires considerable additional infrastructure including Apache Kafka, Redis, MongoDB, and additional MySQL databases, but provides total end-to-end control of the webhook lifecycle. I added a new module for our terraform deployment that creates the required infrastructure and wires it up to the application.

# QA
Deploy this stack using a custom config repo with the value `enable_webhooks` set to `true`. Also the replicated customer you are using need to have "Enable Webhooks" set to true in their customer edit screen in the vendor dashboard. All other frontegg config you see there should be left alone, the default values are correct and shared for all customers. 

# Notable Changes
Along with the webhook additions there are considerable other seamless improvements made to the terraform deployment:
1. Support for specifying existing S3 buckets and VPC. Previously these had to be imported to the terraform state manually.
2. S3 buckets now use versioning.
3. Replicated is now configured automatically to use a self-signed cert with the loadbalancer hostname and a default password for the dashboard is "dozuki". 
4. Part of the automated config is the ability to target a specific replicated sequence number for fresh installs using the variable `replicated_app_sequence_number`. A value of `0` will target the latest release in the channel assigned.
5. The Replicated scheduler has been updated to version 2.52
6. Various terraform providers have been updated to the latest versions. Most notably the AWS and Helm providers.
7. Infracost config and usage estimations have been added to the repo but not yet fully integrated into our Gitops. To run on your local machine, install Infracost and run `INFRACOST_TERRAFORM_BINARY=terragrunt infracost breakdown --usage-file infracost-usage.yml --path ./live/development` from the root of the repo.
